tools/topology2/pipeline.conf

```diff
@@ -0,0 +1,185 @@
+
+#
+# Common pipeline
+#
+# A simple pipeline. All attributes defined herein are namespaced
+# by alsatplg to "pipeline.attribute_name"
+#
+# Usage: this component can be used by declaring in the "pipelines" field of
+# a parent object. i.e.
+#
+# pipelines [
+# 	"pipeline-volume.N" {
+#		format	"s16le"
+#		schedule_period_us	"1000"
+#		schedule_time_domain	"timer"
+#	}
+# ]
+#
+# Where N is a unique integer in the parent object.
+
+
+#
+# Simple pipeline
+#
+# (source) host -> buffer0 -> volume -> buffer1 (sink)
+#
+ClassPipeline."pipeline-volume" {
+
+	############### Boiler Plate Start ####################
+
+	#
+	# Argument used to construct pipeline ID (widget index)
+	#
+	@args [ INDEX ]
+
+	@args.INDEX {
+		type integer
+		default {
+			name pipeline.index
+		}
+	}
+
+	#
+	# Common Attribute Definitions for Pipeline (WIP)
+	#
+
+	# Pipeline ID
+	DefineAttribute."index" {
+		type	"word"
+		# attribute maps to C struct ABI
+	}
+
+	# NO PM flag - not used on DSP or pipeline
+	DefineAttribute."no_pm" {
+		type	"bool"
+		# attribute maps to C struct ABI
+	}
+
+	# Widget Type - Enum in C struct
+	DefineAttribute."type" {
+		type	"string"
+		# attribute maps to C struct ABI
+	}
+
+	############## Boiler Plate End ####################
+
+	#
+	# Bespoke Attribute Definitions for Pipelines (WIP)
+	#
+
+	# Pipeline format
+	DefineAttribute."format" {
+		type	"enum"
+		values [
+			"s16le"
+			"s24le"
+			"s32le"
+		]
+		default "s32le"
+
+		# attribute maps to the following token
+		tuple	"SOF_TKN_PIPE_FORMAT"
+	}
+
+	# Pipeline scheduling period in us
+	DefineAttribute."schedule_period_us" {
+		type	"word"
+		default	"1000"
+
+		# attribute maps to the following token
+		tuple	"SOF_TKN_PIPE_PERIOD"
+	}
+
+	# Pipeline scheduling time domain
+	DefineAttribute."schedule_time_domain" {
+		type 	"word"
+		default "timer"
+
+		# attribute maps to the following token
+		tuple	"SOF_TKN_BUF_PIPE_TIME_DOMAIN"
+	}
+
+
+	#
+	# Attribute constraints (WIP)
+	#
+
+	# Mandatory attributes that all buffers must set
+	attributes."mandatory" [
+		"index"
+		"type"
+		"format"
+		"schedule_period_us"
+		"schedule_time_domain"
+	]
+
+	# optional attributes that some buffers may set
+	attributes."optional" [
+		"TBD"
+	]
+
+	# Immutable attributes that cannnot be changed by parents
+	attributes."immutable" [
+		"type"
+	]
+
+
+	#
+	# Default attribute values - unless overridden if permitted
+	#
+
+	index $INDEX
+	type "pipeline"
+	no_pm "true"
+
+	#
+	# Pipeline Widgets
+	#
+
+	widgets [
+		"host.$INDEX.0" {
+			no_irq	"true"
+		}
+		"buffer.$INDEX.0" {
+			size	"384"
+			caps	"101"
+		}
+		"volume.$INDEX.0" {
+			min_db	"-90"
+			max_db	"0"
+			mute "true"
+		}
+		"buffer.$INDEX.1" {
```

Comment:
The endpoint connections will point to the DMA related buffer. This can then be worked out by compiler.

Reply:
*maybe* this can work, but
1. you're proposing arrays of formats
2. cannot there be cases where a pipeline would decide something different? Do we really want to hard-code that?